browser/ui/webui/brave_rewards_internals_ui.cc

```diff
@@ -0,0 +1,146 @@
+#include "components/prefs/pref_service.h"
+#include "content/public/browser/web_ui.h"
+#include "content/public/browser/web_ui_message_handler.h"
+
+#if !defined(OS_ANDROID)
+#include "brave/components/brave_rewards/resources/grit/brave_rewards_generated_map.h"
+#include "brave/components/brave_rewards/resources/grit/brave_rewards_resources.h"
+#else
+#include "components/brave_rewards/settings/resources/grit/brave_rewards_settings_generated_map.h"
+#include "components/grit/components_resources.h"
+#include "components/grit/components_scaled_resources.h"
+#endif
+
+namespace {
+
+class RewardsInternalsDOMHandler : public content::WebUIMessageHandler {
+ public:
+  RewardsInternalsDOMHandler();
+  ~RewardsInternalsDOMHandler() override;
+
+  void Init();
+
+  // WebUIMessageHandler implementation.
+  void RegisterMessages() override;
+
+ private:
+  bool IsRewardsEnabled() const;
+  void HandleGetRewardsInternalsInfo(const base::ListValue* args);
+  void OnGetRewardsInternalsInfo(
+      std::unique_ptr<brave_rewards::RewardsInternalsInfo> info);
+  void OnPreferenceChanged();
+
+  brave_rewards::RewardsService* rewards_service_;  // NOT OWNED
+  Profile* profile_;
+  std::unique_ptr<PrefChangeRegistrar> pref_change_registrar_;
+  base::WeakPtrFactory<RewardsInternalsDOMHandler> weak_ptr_factory_;
+
+  DISALLOW_COPY_AND_ASSIGN(RewardsInternalsDOMHandler);
+};
+
+RewardsInternalsDOMHandler::RewardsInternalsDOMHandler()
+    : rewards_service_(nullptr), profile_(nullptr), weak_ptr_factory_(this) {}
+
+RewardsInternalsDOMHandler::~RewardsInternalsDOMHandler() {}
+
+void RewardsInternalsDOMHandler::RegisterMessages() {
+  web_ui()->RegisterMessageCallback(
+      "brave_rewards_internals.getRewardsInternalsInfo",
+      base::Bind(&RewardsInternalsDOMHandler::HandleGetRewardsInternalsInfo,
+                 base::Unretained(this)));
+}
+
+void RewardsInternalsDOMHandler::Init() {
+  profile_ = Profile::FromWebUI(web_ui());
+  rewards_service_ =
+      brave_rewards::RewardsServiceFactory::GetForProfile(profile_);
+  PrefService* prefs = profile_->GetPrefs();
+  pref_change_registrar_ = std::make_unique<PrefChangeRegistrar>();
+  pref_change_registrar_->Init(prefs);
+  pref_change_registrar_->Add(
+      brave_rewards::prefs::kBraveRewardsEnabled,
+      base::Bind(&RewardsInternalsDOMHandler::OnPreferenceChanged,
+                 weak_ptr_factory_.GetWeakPtr()));
+}
+
+bool RewardsInternalsDOMHandler::IsRewardsEnabled() const {
+  DCHECK(profile_);
+  return profile_->GetPrefs()->GetBoolean(
+      brave_rewards::prefs::kBraveRewardsEnabled);
+}
+
+void RewardsInternalsDOMHandler::HandleGetRewardsInternalsInfo(
+    const base::ListValue* args) {
+  rewards_service_->GetRewardsInternalsInfo(
+      base::Bind(&RewardsInternalsDOMHandler::OnGetRewardsInternalsInfo,
+                 weak_ptr_factory_.GetWeakPtr()));
+}
+
+void RewardsInternalsDOMHandler::OnPreferenceChanged() {
+  rewards_service_->GetRewardsInternalsInfo(
```

Comment:
do we need to refetch this when `kBraveRewardsEnabled`? Either way I think we should split out `isRewardsEnabled` from `RewardsInternalsInfo`. Looks much better otherwise